kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/serializer/ConfigTokenSerializer.kt

```diff
@@ -0,0 +1,49 @@
+package org.axonframework.extensions.kotlin.serialization.serializer
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.Serializer
+import kotlinx.serialization.builtins.MapSerializer
+import kotlinx.serialization.builtins.serializer
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.descriptors.element
+import kotlinx.serialization.encoding.CompositeDecoder.Companion.DECODE_DONE
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.encoding.decodeStructure
+import kotlinx.serialization.encoding.encodeStructure
+import org.axonframework.eventhandling.tokenstore.ConfigToken
+
+@OptIn(ExperimentalSerializationApi::class)
+@Serializer(forClass = ConfigToken::class)
+class ConfigTokenSerializer : KSerializer<ConfigToken> {
```

Comment:
This is the custom serializer implemented manually for the internal Axon type. These type of implementations are needed if the `@Serializable` annotation is not used in combination with the compiler plugin.

Reply:
I'd assume this should be constructed for every concrete implementation of serializable Axon objects in that case, correct?

This would at least enable moving further with this serializer, so that's good.

Reply:
I just saw https://github.com/Kotlin/kotlinx.serialization/blob/f673bc2/docs/serializers.md#deriving-external-serializer-for-another-kotlin-class-experimental for the first time. This would allow us to skip the 'manual' implementation of the serializer.

Unfortunately I can't yet get it to work, I run into this bug  in the Kotlin complier plugin (https://github.com/Kotlin/kotlinx.serialization/issues/1680).

Reply:
(https://github.com/AxonFramework/extension-kotlin/pull/124#issuecomment-1100655939) I changed the fully-custom example `ConfigTokenSerializer` into a serializer using a `ConfigTokenSurrogate`. Its slightly better maintainable, but still a LOT of work to do (and maintain) for every Axon class.

The work means, for every Axon class:
- Create a Kotlin class with `@Serializable`
- Put all the 
- Manually implement a serializer for that class, using the SerialDescriptor generated by the surrogate.
- During serialization, convert from the real value to a surrogate and serialize it.
- During deserialization, deserialize the surrogate and convert from the surrogate to the value of the real class.

The serializers for surrogates are very simple and similar (only a few lines of code). However the surrogates themselves need to contain exactly the public data fields of the class they are a surrogate for. This will cause bugs if a field is forgotten in the surrogate.